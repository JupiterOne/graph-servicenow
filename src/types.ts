import { IntegrationInstanceConfig } from '@jupiterone/integration-sdk-core';

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  /**
   * Hostname for ServiceNow implementation, e.g. dev94579.service-now.com
   */
  hostname: string;
  /**
   * Parent of all the sys_class_name to ingest
   */
  cmdb_parent: string;
  /**
   * Username for basic auth
   */
  username: string;
  /**
   * Password for basic auth
   */
  password: string;
}
export type CMDBItem = {
  attested_date: string;
  skip_sync: string;
  operational_status: string;
  sys_updated_on: Date;
  attestation_score: string;
  discovery_source: string;
  first_discovered: string;
  sys_updated_by: string;
  due_in: string;
  sys_created_on: Date;
  sys_domain: Asset;
  install_date: Date;
  gl_account: string;
  invoice_number: string;
  sys_created_by: string;
  warranty_expiration: string;
  asset_tag: string;
  fqdn: string;
  change_control: string;
  owned_by: string;
  checked_out: string;
  sys_domain_path: string;
  business_unit: string;
  delivery_date: string;
  maintenance_schedule: string;
  install_status: string;
  cost_center: Asset;
  attested_by: string;
  supported_by: string;
  dns_domain: string;
  name: string;
  assigned: Date;
  life_cycle_stage: string;
  purchase_date: string;
  subcategory: string;
  short_description: string;
  assignment_group: string;
  managed_by: string;
  managed_by_group: string;
  can_print: string;
  last_discovered: string;
  sys_class_name: string;
  manufacturer: Asset;
  sys_id: string;
  po_number: string;
  checked_in: string;
  sys_class_path: string;
  life_cycle_stage_status: string;
  mac_address: string;
  vendor: Asset;
  company: Asset;
  justification: string;
  model_number: string;
  department: Asset;
  assigned_to: Asset;
  start_date: string;
  comments: string;
  cost: string;
  attestation_status: string;
  sys_mod_count: string;
  monitor: string;
  serial_number: string;
  ip_address: string;
  model_id: Asset;
  duplicate_of: string;
  sys_tags: string;
  cost_cc: string;
  order_date: string;
  schedule: string;
  support_group: string;
  environment: string;
  due: string;
  attested: string;
  correlation_id: string;
  unverified: string;
  attributes: string;
  location: Asset;
  asset: Asset;
  category: string;
  fault_count: string;
  lease_id: string;
};

export type Asset = {
  link: string;
  value: string;
};
export type DictionaryItem = {
  super_class: Asset;
  name: string;
};
