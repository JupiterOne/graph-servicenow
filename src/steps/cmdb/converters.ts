import {
  Entity,
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { Entities } from '../../constants';
import { CMDBItem } from '../../types';
import { skippedRawDataSource } from '../../util/graphObject';

export function createCMDBEntity(
  data: CMDBItem,
  sysClassNames: string[],
): Entity {
  //For now all sysclassnames use the same converter. We can always create different converters.
  const customFields: { [key: string]: string } = {};
  for (const [key, value] of Object.entries(data)) {
    //custom fields
    if (key.startsWith('u_') || key.includes('_u_')) {
      const customFieldsKey = key.split('_').join('-');

      // Don't include undefined/null values
      if (value === undefined || value === null) {
        continue;
      }

      let finalValue = value;
      if (typeof value == 'object') {
        // Only stringify if value is an object to avoid double quotes around primitive values.
        finalValue = JSON.stringify(value);
      } else {
        finalValue = value;
      }

      customFields[customFieldsKey] = finalValue;
    }
  }

  return createIntegrationEntity({
    entityData: {
      source: skippedRawDataSource, // Raw data is really big.
      assign: {
        _class: Entities.CMDB_OBJECT._class,
        _type: Entities.CMDB_OBJECT._type,
        _key: data.sys_id,
        id: data.sys_id,
        correlationId: data.correlation_id,
        name: data.name,
        displayName: data.name,
        sysClassNames: sysClassNames,
        installStatus: data.install_status,
        attributes: data.attributes,
        businessUnit: data.business_unit,
        category: data.category,
        environment: data.environment,
        ipAddress: data.ip_address,
        macAddress: data.mac_address,
        operationalStatus: data.operational_status,
        serial: data.serial_number,
        subcategory: data.subcategory,
        warrantyExpiration: parseTimePropertyValue(data.warranty_expiration),
        model: data.model_number,
        managedBy: data.managed_by?.value,
        ownedBy: data.owned_by?.value,
        managedByGroup: data.managed_by_group?.value,
        assignedTo: data.assigned_to?.value,
        createdOn: parseTimePropertyValue(data.sys_created_on + ' UTC'),
        updatedOn: parseTimePropertyValue(data.sys_updated_on + ' UTC'),
        ...customFields,
      },
    },
  });
}
