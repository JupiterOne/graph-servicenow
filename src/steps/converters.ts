import {
  Entity,
  createIntegrationEntity,
  Relationship,
  createDirectRelationship,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';
import { Entities } from '../constants';

export function getUserKey(id: string, name: string): string {
  return `service_now_user:${id}:${name}`;
}

export function getGroupKey(id: string, name: string): string {
  return `service_now_group:${id}:${name}`;
}

export function getIncidentKey(id: string, name: string): string {
  return `service_now_incident:${id}:${name}`;
}

function convertCommonServiceNowProperties(
  serviceNowObject: any,
): Record<string, string | boolean | number | null | undefined> {
  return {
    createdOn: new Date(serviceNowObject.sys_created_on + ' UTC').valueOf(),
    updatedOn: new Date(serviceNowObject.sys_updated_on + ' UTC').valueOf(),
    id: serviceNowObject.sys_id,
  };
}

export function createAccountEntity(hostname: string): Entity {
  return createIntegrationEntity({
    entityData: {
      source: {},
      assign: {
        _class: Entities.ACCOUNT._class,
        _type: Entities.ACCOUNT._type,
        _key: hostname,
        name: hostname,
        displayName: hostname,
      },
    },
  });
}

export function createUserEntity(user: any): Entity {
  delete user.user_password;
  return createIntegrationEntity({
    entityData: {
      source: user,
      assign: {
        ...convertCommonServiceNowProperties(user),
        _class: Entities.USER._class,
        _type: Entities.USER._type,
        _key: getUserKey(user.sys_id, user.user_name),
        username: user.user_name,
        active: user.active === 'true',
        email: user.email ? user.email : undefined,
      },
    },
  });
}

export function createGroupEntity(group: any): Entity {
  return createIntegrationEntity({
    entityData: {
      source: group,
      assign: {
        ...convertCommonServiceNowProperties(group),
        _class: Entities.GROUP._class,
        _type: Entities.GROUP._type,
        _key: getGroupKey(group.sys_id, group.email),
        active: group.active === 'true',
        email: group.email ? group.email : undefined,
      },
    },
  });
}

export function createIncidentEntity(incident: any): Entity {
  return createIntegrationEntity({
    entityData: {
      source: incident,
      assign: {
        ...convertCommonServiceNowProperties(incident),
        _class: Entities.INCIDENT._class,
        _type: Entities.INCIDENT._type,
        _key: getIncidentKey(incident.sys_id, incident.number),
        name: incident.number,
        displayName: incident.number,
        severity: incident.severity,
        category: incident.category,
        reporter: incident.opened_by.value,
        impact: incident.impact,
        resolvedAt: incident.resolved_at,
        active: incident.active === 'true',
        reportable: false,
      },
    },
  });
}

export function createIncidentAssigneeRelationship(
  incident: any,
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.ASSIGNED,
    fromType: Entities.INCIDENT._type,
    fromKey: incident.sys_id,
    toType: Entities.USER._type,
    toKey: incident.assigned_to.value,
  });
}

export function createGroupGroupRelationship(
  groupEntity: Entity,
  parent: { link: string; value: string },
): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    fromType: Entities.GROUP._type,
    fromKey: parent.value,
    toType: Entities.GROUP._type,
    toKey: groupEntity._key,
  });
}

export function createGroupUserRelationship(groupUser: any): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    fromType: Entities.GROUP._type,
    fromKey: groupUser.group.value,
    toType: Entities.USER._type,
    toKey: groupUser.user.value,
    properties: convertCommonServiceNowProperties(groupUser),
  });
}
