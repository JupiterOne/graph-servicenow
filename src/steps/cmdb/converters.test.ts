import { CMDBItem } from '../../types';
import { createCMDBEntity } from './converters';

describe('createCMDBEntity', () => {
  test('that properties are converted correctly', () => {
    const data: CMDBItem = ({
      sys_id: 'test-id',
      sys_created_on: '2019-04-05 21:09:12',
      sys_updated_on: '2020-04-05 21:09:12',
      name: 'test-name',
      u_custom_1: 'custom 1',
      u_custom_2: '',
      test_u_custom_3: 'custom 3',
      u_custom_4: undefined,
      u_custom_object: {
        a: 'a',
        b: 'b',
        c: 123,
      },
      u_custom_number: 234,
      u_custom_boolean: true,
    } as unknown) as CMDBItem;

    const entity = createCMDBEntity(data, ['fred', 'george']);
    expect(entity).toMatchSnapshot();
  });
});
