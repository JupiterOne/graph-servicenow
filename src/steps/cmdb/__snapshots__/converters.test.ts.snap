// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCMDBEntity that properties are converted correctly 1`] = `
{
  "_class": [
    "Configuration",
  ],
  "_key": "test-id",
  "_rawData": [
    {
      "name": "default",
      "rawData": {
        "uploadStatus": "SKIPPED",
        "uploadStatusReason": "Raw data currently disabled for this entity type",
      },
    },
  ],
  "_type": "service_now_cmdb_object",
  "assignedTo": undefined,
  "attributes": undefined,
  "businessUnit": undefined,
  "category": undefined,
  "correlationId": undefined,
  "createdOn": 1554498552000,
  "displayName": "test-name",
  "environment": undefined,
  "id": "test-id",
  "installStatus": undefined,
  "ipAddress": undefined,
  "macAddress": undefined,
  "managedBy": undefined,
  "managedByGroup": undefined,
  "model": undefined,
  "name": "test-name",
  "operationalStatus": undefined,
  "ownedBy": undefined,
  "serial": undefined,
  "subcategory": undefined,
  "sysClassNames": [
    "fred",
    "george",
  ],
  "testUCustom3": "custom 3",
  "uCustom1": "custom 1",
  "uCustom2": "",
  "uCustomBoolean": true,
  "uCustomNumber": 234,
  "uCustomObject": "{"a":"a","b":"b","c":123}",
  "updatedOn": 1586120952000,
  "warrantyExpiration": undefined,
}
`;
